README.md
# About

The mpt-status software is a query tool to access the running configuration and
status of LSI SCSI HBAs. This is a completely rewritten version of the original
mpt-status-1.0 tool written by Matt Braithwaite. mpt-status allows you to
monitor the health and status of your RAID setup.

Currently supported and tested HBAs are:

LSI 1030 SCSI RAID storage controller
LSI SAS1064 SCSI RAID storage controller
LSI SAS1068 SCSI RAID storage controller
LSI SAS 3442-R SCSI RAID storage controller

Since the tool is using the MPI (message passing interface) chances are high
that the basic information regarding RAID status will be available for all
LSI based controllers. Just give it a try and report back.


# Requirements

You should have mptbase and mptctl loaded or compiled into the kernel.


# Reported working hardware configuration

Sun Fire X4100
Sun Fire X4200
Sun Fire V20z
Sun Fire V40z
Dell PE2600
Intel Server with SE7520BD2S boards
HP ProLiant DL320 G4
IBM eServer BladeCenter LS20


# Where can I get this fine piece of software?

Homepage of maintained version: http://github.com/baruch/mpt-status

The original version of this tool is not maintained anymore by its author. He
was kind enough to put a link to the current mpt-status home. Thanks, Matt.

The second maintainer was Roberto Nibali, whose site disappeared at some point.

# Distro related support

To my avail following distributions have mpt-status included:

| Distro | Status    | URL |
| ------ | --------- | --- |
| Debian | OK        | http://packages.debian.org/unstable/admin/mpt-status |
| Suse	 | NOK (old) | http://www.novell.com/products/linuxpackages/professional/mpt-status.html |
|	 | NOK (old) | http://www.novell.com/products/linuxpackages/enterpriseserver/SP3/ia64/mpt-status.html |

Support for following distros can be found through google and rpm searches 
(status unknown):

    PLD 
    Turbolinux

I have provided spec (for RPM) files for OpenSuse and Red Hat, however they are
mostly untested. Have a look into the contrib sub directory.


# How can I compile, install or uninstall mpt-status?

Read doc/INSTALL for further information on building mpt-status. Basically you
should be fine invoking

    make

A simple (as root)

    make install install_doc

should install the binary to $PREFIX/sbin, and can be uninstalled as follows:

    make uninstall


# What does this software provide you?

For now it should be improved regarding compilation on different systems, also
different kernels. But the output has changed as well. Example:

    SR2400:~# ./mpt-status -i 1
    ioc0 vol_id 1 type IM, 2 phy, 68 GB, state OPTIMAL, flags ENABLED
    ioc0 phy 0 scsi_id 1 SEAGATE  ST373207LC       0003, 68 GB, state ONLINE, flags NONE
    ioc0 phy 1 scsi_id 3 SEAGATE  ST373207LC       0003, 68 GB, state ONLINE, flags NONE

    [root@redhatAS-3-U4-X86_64 mpt-status-1.1.5-new]# ./mpt-status

    You seem to have no SCSI disks attached to your HBA or you have
    them on a different scsi_id. To get your SCSI id, run:

    mpt-status -p

    [root@redhatAS-3-U4-X86_64 mpt-status-1.1.5-new]# ./mpt-status -p
    Found SCSI id=2, use ''mpt-status -i 2`` to get more information.

    [root@redhatAS-3-U4-X86_64 mpt-status-1.1.5-new]# ./mpt-status -i 2
    ioc0 vol_id 2 type IM, 2 phy, 67 GB, state OPTIMAL, flags ENABLED
    ioc0 phy 1 scsi_id 4 FUJITSU  MAV2073RCSUN72G  0301, 68 GB, state ONLINE, flags NONE
    ioc0 phy 0 scsi_id 3 FUJITSU  MAV2073RCSUN72G  0301, 68 GB, state ONLINE, flags NONE

    [root@redhatAS-3-U4-X86_64 mpt-status-1.1.5-new]# ./mpt-status -i 2 -s
    vol_id 2 OPTIMAL
    phys_id 1 ONLINE
    phys_id 0 ONLINE

A new style output has made it into the sources, however due to backwards
compatibility reasons it's not on per default. You can enable the new style
output by invoking mpt-status as follows:

    mpt-status --newstyle

There is also a debug mode since the mpt-status-1.2.0 release.


# Hardware, Software and Distribution Compatibility

The mpt-status software is known and reported to compile and work on following
Linux distributions:

* SuSE		: 9.0, 9.1, 9.2, 9.3, 10.0, 10.1, 10.2, SLES8, SLES9, SLES10
* Redhat	: 7.3, 9.0, RHEL3, RHEL4, FC1, FC2, FC3, FC4, FC5
* Debian	: all so far
* Ubuntu	: edgy
* Gentoo	: 2005.x
* ulibc-based	: all

Kernels:
Linux: 2.4.x, 2.6.x

Architectures:
* i386
* x86\_64


# List of Hardware configuration

This list should eventually be exported somewhere so decision makers and 
hardware buyers can easily find out if their hardware is a) supported and
b) can be monitored.

If you have successfully running mpt-status on a platform not mentioned below,
please send me following output:

* dmesg -s 1000000 (best after having rebooted your machine)
* lspci -v
* dmidecode


doc/Changelog
2006-11-01, ratz [ratz@drugphish.ch]:
	o man/mpt-status.8
		- merged contributions from Jean Delvare
		- updated with missing sections
	o doc/*
		- merged typo/spelling fixes from Jean Delvare
		- tidy up 64bit information since this problem is solved now
	o mpt-status.c
		- removed MPT_STATUS_DEBUG, since this was an unfinished
		  duplicate version of --debug.

2006-10-26, ratz [ratz@drugphish.ch]:
	o man/{README,mpt-status.8}
		- added man page written by Steffen Joeris for the Debian
		  project.
	o Makefile
		- integrated preliminary support for the man page installation
		  and uninstallation through make install, resp. make uninstall
		- fixed once again the KERNEL_PATH stuff after Jean Delvare
		  from Suse mentioned that it broke Suse builds. This time it
		  should work for both, Suse and Redhat (Fedora).
	o mpt-status.c
		- added detection if no RAID is configured on HBA. This only
		  works with the "--newstyle" or "-n" option for safety reasons
		  and has not yet been tested.

2006-10-25, ratz [ratz@drugphish.ch]:
	o Makefile
		- fixed brokenness due to previously applied patch from Rich :)
		  The point is that LSI-related headers are in the source and
		  not the include directory of the build symlink.
	o mpt-status.c
		- Suse packages are compiled with additional flags to detect
		  possibly problematic code. The checker complained about a call
		  to system() with unchecked return value in mpt-status. Applied
		  patch from Jean Delvare.

2006-06-29, ratz [ratz@drugphish.ch]:
	o Makefile
		- applied patch from Rich Edelman to set KERNEL_PATH more
		  intelligently 
	o contrib/contrib/mpt-status.spec
		- generic spec file from Rich Edelman with patch support
		- changed quite a lot again actually regarding wording
		  and also removed the patch section
	o contrib/*diff
		- remove obsolete diff from tree regarding sanitation

2006-04-20-
2006-05-02, ratz [ratz@drugphish.ch]:
	* Version: 1.2.0-RC7 (not released), after 3 almost complete rewrites
		   too many changes happened in between the last two weeks to
		   list them properly. Kernel header sanitizing might not be
		   possible after all.
	o mpt-status.c
		- proper support for 64-bit nodes
		- added support for spare disk information
		- added preliminary support for the sync percentage information
	o doc/TODO
		- updated

2006-04-20, ratz [ratz@drugphish.ch]:
	o mpt-status.c
		- reworked physical disk section
		- integrated the exit status patch submitted by Leandro Santi

2006-04-16, ratz [ratz@drugphish.ch]:
	o mpt-status.c
		- merged Steffen Joeris patch to support integrating mpt-status
		  into the KFreeBSD project.

2006-03-09, ratz [ratz@drugphish.ch]:
	o mpt-status.c
		- Martin Hamant reported an issue regarding the IOC unit, which
		  did not work for more than 1 controller. Fix it, by setting
		  ioc_unit as part of the address unit passed over to readpage.

2006-03-06, ratz [ratz@drugphish.ch]:
	o mpt-status.c
		- removed SGE debugging lines
		- added checkForLibraryMode() which reads the __MPT_STATUS_LIB
		  environment variable and if set will initialize the library
		  functionality of mpt-status.
		- changed the output to please Martin Hamant and to also ease
		  up the scriptability.

2006-03-05, ratz [ratz@drugphish.ch]:
	o doc/Changelog
		- fixed overly long line
	o mpt-status.c
		- added human readable output for IM, IME, IS, which is nothing
		  else than RAID0, RAID1E and RAID1.

2006-03-04, ratz [ratz@drugphish.ch]:
	* Version: 1.1.5 [released] (Codename: Sebnem)
	o doc/README
		- major overhaul regarding distribution availability and
		  installation process
	o doc/FAQ
		- added a possible FAQ document containing parts of an email
		  I wrote to explain the output of mpt-status
	o doc/ReleaseNotes
		- wrote Problems, Upgrade, Outlook sections
	o incl/sles.h
		- added preliminary work to kernel header sanitation done by
		  SuSE
	o mpt-status.c
		- made the struct divisior 32/64 bit safe
		- removed irritating compile warning for type signedness since
		  too many people complained about it and before the kernel
		  headers are not sanitized, there is no point in trying to
		  address this deficiency
		- preliminary work done for IOCpage5_t support (hotspare)
		- fixed missing signedness casts
	o contrib/sizes.c
		- added a little tool to check the machine's type sizes
	o contrib/*.spec*
		- added various untested spec files

2006-02-11, ratz [ratz@drugphish.ch]:
	o Makefile
		- Added support for DESTDIR, patch kindly submitted by Petter
		  Reinholdtsen

2006-01-22, ratz [ratz@drugphish.ch]:
	o mpt-status.{ch}
		- s/log_id/vol_id/ in the first line if -s is submitted
		- added --probe_id which will probe for the first scsi id
		- fixed a long standing bug resulting in a coredump when using
		  the -d parameter
		- moved print_status_only, id_of_primary_device, ioc_unit into
		  global context, so there is no need to pass those variables
		  through the stack all the time
		- reorganized long_options and usage to be in alphabetical
		  order
		- fixed read_page to actually work on LSI-SAS1064 controllers
	o doc/*
		- major cleanup and improved documentation

2005-11-14, ratz [ratz@drugphish.ch]:
	o mpt-status.c
		- removed Nietsche, added "--autoload" which will try to exec
		  /sbin/modprobe mptctl.

2005-11-12, ratz [ratz@drugphish.ch]:
	* Version: 1.1.4 [released]

2005-11-11, ratz [ratz@drugphish.ch]:
	o doc/DeveloperNotes
		- Added information from the mpt(4) man page of OpenBSD
	o mpt-status.h
		- Fixed VERSION string
	o mpt-status.c
		- Merged patch by Thomas Richer to allow setting the ioc unit
		- Added Copyright note
		- Removed the modprobe call since this is braindead. Not all
		  people have mpt loaded as module. Making it easy to use does
		  not mean making it dumb.
		- Rewrote the primitive parameter passing on command line to
		  more sophisticated GNU style option parsing.
		- Added some more switches, check them out with --help

mpt-status.h
#ifndef _MPT_STATUS_H
#define _MPT_STATUS_H

#include <sys/ioctl.h>
#include <stdint.h>

#define __user
typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#include "includes/mpi_type.h"
#include "includes/mpi.h"
#include "includes/mpi_cnfg.h"
#include "includes/mpi_ioc.h"
#include "includes/mpi_raid.h"
#include "includes/mptctl.h"

#define VERSION "1.2.0"

#define BIG 1024
#define REALLYBIG 10240
#define PRINT_STATUS_ONLY 0

#define REPLY_SIZE 128
#define MPT_EXIT_OKAY		(0)
#define MPT_EXIT_VOL_OPTIMAL	(0)
#define MPT_EXIT_VOL_FAILED	(1 << 1)
#define MPT_EXIT_VOL_DEGRADED	(1 << 2)
#define MPT_EXIT_VOL_RESYNC	(1 << 3)
#define MPT_EXIT_PHYSDISK_ERROR	(1 << 4)
#define MPT_EXIT_PHYSDISK_WARN	(1 << 5)
#define MPT_EXIT_UNKNOWN	(1 << 0)

#define MPT_EXIT_NOCLOSE	(1 << 10)
#define MPT_EXIT_FREEMEM	(1 << 11)

#define DATA_DIR_NONE           0
#define DATA_DIR_IN             1
#define DATA_DIR_OUT            2

#define MPT_FLAGS_FREE_MEM      0x00
#define MPT_FLAGS_KEEP_MEM      0x01
#define MPT_FLAGS_DUMP_REPLY    0x02
#define MPT_FLAGS_DUMP_DATA     0x04

static char *wrong_scsi_id =
"\nYou seem to have no SCSI disks attached to your HBA or you have\n"
"them on a different scsi_id. To get your SCSI id, run:\n\n"
"    mpt-status -p\n";

typedef struct _SYNC_INFO {
	int	blocks_total;
	int	blocks_done;
	int	blocks_left;
} SyncInfoData;

#endif /* End of mpt-status.h */


incl/mpt-sanitized.h
#define MPI_FUNCTION_CONFIG				(0x04)
#define MPI_CONFIG_ACTION_PAGE_HEADER			(0x00)
#define MPI_CONFIG_PAGETYPE_RAID_VOLUME			(0x08)
#define MPI_CONFIG_ACTION_PAGE_READ_CURRENT		(0x01)
#define MPI_RAIDVOLPAGE0_PAGEVERSION			(0x01)
#define MPI_RAIDVOL0_STATUS_FLAG_ENABLED		(0x01)
#define MPI_RAIDVOL0_STATUS_FLAG_QUIESCED		(0x02)
#define MPI_RAIDVOL0_STATUS_FLAG_RESYNC_IN_PROGRESS	(0x04)
#define MPI_RAIDVOL0_STATUS_FLAG_VOLUME_INACTIVE	(0x08)
#define MPI_RAIDVOL0_STATUS_STATE_OPTIMAL		(0x00)
#define MPI_RAIDVOL0_STATUS_STATE_DEGRADED		(0x01)
#define MPI_RAIDVOL0_STATUS_STATE_FAILED		(0x02)
#define MPI_CONFIG_PAGETYPE_RAID_PHYSDISK		(0x0A)
#define MPI_RAIDPHYSDISKPAGE0_PAGEVERSION		(0x00)
#define MPI_PHYSDISK0_STATUS_FLAG_OUT_OF_SYNC		(0x01)
#define MPI_PHYSDISK0_STATUS_FLAG_QUIESCED		(0x02)
#define MPI_PHYSDISK0_STATUS_ONLINE			(0x00)
#define MPI_PHYSDISK0_STATUS_MISSING			(0x01)
#define MPI_PHYSDISK0_STATUS_NOT_COMPATIBLE		(0x02)
#define MPI_PHYSDISK0_STATUS_FAILED			(0x03)
#define MPI_PHYSDISK0_STATUS_FAILED_REQUESTED		(0x06)
#define MPI_PHYSDISK0_STATUS_INITIALIZING		(0x04)
#define MPI_PHYSDISK0_STATUS_OFFLINE_REQUESTED		(0x05)
#define MPI_PHYSDISK0_STATUS_FAILED_REQUESTED		(0x06)
#define MPI_PHYSDISK0_STATUS_OTHER_OFFLINE		(0xFF)
#define MPI_RAID_VOL_PAGE_0_PHYSDISK_MAX		(1)

#define MPT_MAGIC_NUMBER	'm'
#define MPTCOMMAND		_IOWR(MPT_MAGIC_NUMBER,20,struct mpt_ioctl_command)

typedef struct ConfigPageHeader_t {
	uint8_t PageVersion;	/* 00h */
	uint8_t PageLength;	/* 01h */
	uint8_t PageNumber;	/* 02h */
	uint8_t PageType;	/* 03h */
} ConfigPageHeader_t;

typedef struct SGESimpleUnion_t {
	uint32_t FlagsLength;
	union {
		uint32_t Address32;
		uint64_t Address64;
	} u;
} SGESimpleUnion_t;

typedef struct SGEChainUnion_t {
	uint16_t Length;
	uint8_t NextChainOffset;
	uint8_t Flags;
	union {
		uint32_t Address32;
		uint64_t Address64;
	} u;
} SGEChainUnion_t;

typedef struct SGEIOUnion_t {
	union {
		SGESimpleUnion_t Simple;
		SGEChainUnion_t Chain;
	} u;
} SGEIOUnion_t;

typedef struct Config_t {
	uint8_t Action;		/* 00h */
	uint8_t Reserved;	/* 01h */
	uint8_t ChainOffset;	/* 02h */
	uint8_t Function;	/* 03h */
	uint16_t ExtPageLength;	/* 04h */
	uint8_t ExtPageType;	/* 06h */
	uint8_t MsgFlags;	/* 07h */
	uint32_t MsgContext;	/* 08h */
	uint8_t Reserved2[8];	/* 0Ch */
	ConfigPageHeader_t Header;	/* 14h */
	uint32_t PageAddress;	/* 18h */
	SGEIOUnion_t PageBufferSGE;	/* 1Ch */
} Config_t;

typedef struct ConfigReply_t {
	uint8_t Action;		/* 00h */
	uint8_t Reserved;	/* 01h */
	uint8_t MsgLength;	/* 02h */
	uint8_t Function;	/* 03h */
	uint16_t ExtPageLength;	/* 04h */
	uint8_t ExtPageType;	/* 06h */
	uint8_t MsgFlags;	/* 07h */
	uint32_t MsgContext;	/* 08h */
	uint8_t Reserved2[2];	/* 0Ch */
	uint16_t IOCStatus;	/* 0Eh */
	uint32_t IOCLogInfo;	/* 10h */
	ConfigPageHeader_t Header;	/* 14h */
} ConfigReply_t;

typedef struct RaidVol0Status_t {
	uint8_t Flags;		/* 00h */
	uint8_t State;		/* 01h */
	uint16_t Reserved;	/* 02h */
} RaidVol0Status_t;

typedef struct RaidVol0Settings {
	uint16_t Settings;	/* 00h */
	uint8_t HotSparePool;	/* 01h *//* MPI_RAID_HOT_SPARE_POOL_ */
	uint8_t Reserved;	/* 02h */
} RaidVol0Settings;

typedef struct RaidVol0PhysDisk_t {
	uint16_t Reserved;	/* 00h */
	uint8_t PhysDiskMap;	/* 02h */
	uint8_t PhysDiskNum;	/* 03h */
} RaidVol0PhysDisk_t;

typedef struct RaidVolumePage0_t {
	ConfigPageHeader_t Header;	/* 00h */
	uint8_t VolumeID;	/* 04h */
	uint8_t VolumeBus;	/* 05h */
	uint8_t VolumeIOC;	/* 06h */
	uint8_t VolumeType;	/* 07h *//* MPI_RAID_VOL_TYPE_ */
	RaidVol0Status_t VolumeStatus;	/* 08h */
	RaidVol0Settings VolumeSettings;	/* 0Ch */
	uint32_t MaxLBA;	/* 10h */
	uint32_t Reserved1;	/* 14h */
	uint32_t StripeSize;	/* 18h */
	uint32_t Reserved2;	/* 1Ch */
	uint32_t Reserved3;	/* 20h */
	uint8_t NumPhysDisks;	/* 24h */
	uint8_t Reserved4;	/* 25h */
	uint16_t Reserved5;	/* 26h */
	RaidVol0PhysDisk_t PhysDisk[MPI_RAID_VOL_PAGE_0_PHYSDISK_MAX];	/* 28h */
} RaidVolumePage0_t;

typedef struct RaidPhysDiskSettings_t {
	uint8_t SepID;		/* 00h */
	uint8_t SepBus;		/* 01h */
	uint8_t HotSparePool;	/* 02h *//* MPI_RAID_HOT_SPARE_POOL_ */
	uint8_t PhysDiskSettings;	/* 03h */
} RaidPhysDiskSettings_t;

typedef struct RaidPhysDisk0InquiryData {
	uint8_t VendorID[8];	/* 00h */
	uint8_t ProductID[16];	/* 08h */
	uint8_t ProductRevLevel[4];	/* 18h */
	uint8_t Info[32];	/* 1Ch */
} RaidPhysDisk0InquiryData;

typedef struct RaidPhysDiskStatus_t {
	uint8_t Flags;		/* 00h */
	uint8_t State;		/* 01h */
	uint16_t Reserved;	/* 02h */
} RaidPhysDiskStatus_t;

typedef struct RaidPhysDisk0ErrorData_t {
	uint8_t ErrorCdbByte;	/* 00h */
	uint8_t ErrorSenseKey;	/* 01h */
	uint16_t Reserved;	/* 02h */
	uint16_t ErrorCount;	/* 04h */
	uint8_t ErrorASC;	/* 06h */
	uint8_t ErrorASCQ;	/* 07h */
	uint16_t SmartCount;	/* 08h */
	uint8_t SmartASC;	/* 0Ah */
	uint8_t SmartASCQ;	/* 0Bh */
} RaidPhysDisk0ErrorData_t;

typedef struct RaidPhysDiskPage0_t {
	ConfigPageHeader_t Header;	/* 00h */
	uint8_t PhysDiskID;	/* 04h */
	uint8_t PhysDiskBus;	/* 05h */
	uint8_t PhysDiskIOC;	/* 06h */
	uint8_t PhysDiskNum;	/* 07h */
	RaidPhysDiskSettings_t PhysDiskSettings;	/* 08h */
	uint32_t Reserved1;	/* 0Ch */
	uint32_t Reserved2;	/* 10h */
	uint32_t Reserved3;	/* 14h */
	uint8_t DiskIdentifier[16];	/* 18h */
	RaidPhysDisk0InquiryData InquiryData;	/* 28h */
	RaidPhysDiskStatus_t PhysDiskStatus;	/* 64h */
	uint32_t MaxLBA;	/* 68h */
	RaidPhysDisk0ErrorData_t ErrorData;	/* 6Ch */
} RaidPhysDiskPage0_t;

typedef struct mpt_ioctl_header {
	unsigned int iocnum;	/* IOC unit number */
	unsigned int port;	/* IOC port number */
	int maxDataSize;	/* Maximum Num. bytes to transfer on read */
} mpt_ioctl_header;

struct mpt_ioctl_command {
	mpt_ioctl_header hdr;
	int timeout;		/* optional (seconds) */
	char *replyFrameBufPtr;
	char *dataInBufPtr;
	char *dataOutBufPtr;
	char *senseDataPtr;
	int maxReplyBytes;
	int dataInSize;
	int dataOutSize;
	int maxSenseBytes;
	int dataSgeOffset;
	char MF[1];
};
